pennylane/beta/interfaces/autograd.py

```diff
@@ -100,7 +101,7 @@ def _update_trainable_params(self):
         trainable_params = set()
 
         for idx, p in enumerate(params):
-            if getattr(p, "requires_grad", True):
+            if getattr(p, "requires_grad", False) or isinstance(p, ArrayBox):

```

Comment:
On the forward pass, the user must specify `requires_grad=True` for parameters to be marked as default. Previously, arrays without this attribute, and all python numeric literals, would by default be treated as differentiable.

On the backward pass, we count the existence of an arraybox as differentiable.

@trbromley, this _might_ reduce the number of autograd computations?

Reply:
Does that mean we now have to run a circuit forward before optimising it, like in pytorch? In other words, what makes the backward pass have the `ArrayBox`?

Reply:
> Does that mean we now have to run a circuit forward before optimising it

Nope, it works perfectly fine simply defining the qnode, creating the grad function, and calling the grad function straight away.

>  what makes the backward pass have the ArrayBox?

I'm not entirely sure, just something I've noticed with how autograd works.

* When you call an Autograd function, it always evaluates it numerically.

* When you call the `grad_fn` of an autograd function, it evaluates the function with `ArrayBox` objects, and then back-propagates to construct the final gradient. I _think_ the presence of `ArrayBox`'s is to help with gradient accumulation? But I'm not entirely sure.

Reply:
`getattr(p, "requires_grad", True)` would previously have the logic:

* If `isinstance(p, qml.numpy.tensor)`, then it would simply use the specified version of `p.requires_grad`.

* If `not isinstance(p, qml.numpy.tensor)` (for example, if `p` was a float or a list, and so didn't have a `requires_grad` attribute), the tape would previously continue to assume it was a trainable parameter.

The change here only affects the second bullet point. With this change, non-PL numpy arrays, including floats, tuples, lists, etc., are now assumed to be constant by default. If you want to make a float trainable, you _have_ to now do `np.array(0.564, requires_grad=True)`.

Note that the above applies only to the (autograd) tape directly, the (autograd) QNode has different logic.

Reply:
Ohhhhh! I was reading it as `getattr(p, "requires_grad") is False` :facepalm: Ok now I agree it looks good! Thanks.